demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1354 @@
+
+    if( returnStatus == EXIT_SUCCESS )
+    {
+        /* Wait till OTA library is stopped, output statistics for currently running
+         * OTA job */
+        while( ( ( state = OTA_GetState() ) != OtaAgentStateStopped ) )
+        {
+            if( mqttSessionEstablished != true )
+            {
+                /* Connect to MQTT broker and create MQTT connection. */
+                returnStatus = establishConnection();
+
+                if( returnStatus == EXIT_SUCCESS )
+                {
+                    mqttSessionEstablished = true;
+
+                    /* Check if OTA process was suspended and resume if required. */
+                    if( state == OtaAgentStateSuspended )
+                    {
+                        /* Resume OTA operations. */
+                        OTA_Resume();
+                    }
+                    else
+                    {
+                        /* Send start event to OTA Agent.*/
+                        eventMsg.eventId = OtaAgentEventStart;
+                        OTA_SignalEvent( &eventMsg );
+                    }
+                }
+            }
+
+            if( mqttSessionEstablished == true )
+            {
+                /* Loop to receive packet from transport interface. */
+                mqttStatus = MQTT_ProcessLoop( &mqttContext, OTA_EXAMPLE_TASK_DELAY_MS );
+
+                /* Get OTA statistics for currently executing job. */
+                OTA_GetStatistics( &otaStatistics );
+
+                if( mqttStatus == MQTTSuccess )
+                {
+                    LogInfo( ( " Received: %u   Queued: %u   Processed: %u   Dropped: %u",
+                               otaStatistics.otaPacketsReceived,
+                               otaStatistics.otaPacketsQueued,
+                               otaStatistics.otaPacketsProcessed,
+                               otaStatistics.otaPacketsDropped ) );
+                }
+                else
+                {
+                    LogError( ( "MQTT_ProcessLoop returned with status = %u.",
+                                mqttStatus ) );
+
+                    /* Disconnect from broker and close connection. */
+                    disconnect();
+
+                    /* Set connection flag to false. */
+                    mqttSessionEstablished = false;
+
+                    /* Suspend OTA operations. */
+                    otaRet = OTA_Suspend();
+
+                    if( otaRet != OtaErrNone )
+                    {
+                        LogError( ( "OTA failed to suspend. "
+                                    "StatusCode=%d.", otaRet ) );
+                    }
+                    else
+                    {
+                        while( ( state = OTA_GetState() ) != OtaAgentStateSuspended )
+                        {
+                            /* Wait for OTA Library state to suspend */
+                            sleep( OTA_EXAMPLE_TASK_DELAY_MS );
+                        }
```

Comment:
Can this possibly get into an infinite loop? Is there an upper limit on the time required for the OTA agent to be suspended after which it can be treated as an error?

Reply:
Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1418/commits/7bed0e4558dfd4a6c985fc6da34260501e76d4c7